Add FixAll in containing type/member support for code fixes

mavasani opened the pull request:
Currently, one can apply FixAll operations to Document, Project and Solution scopes. Now, we also support fixing all diagnostics in containing member and type scopes, wherever applicable.

Closes #60029 

## Screenshot

![FixAllInContainingMemberAndType](https://user-images.githubusercontent.com/10605811/156998125-30d0ff30-ef01-4ab7-bcf7-90cfa15a7015.gif)

## Note

1. Fix all in `ContainingType` scope fixes all diagnostics in _all the partial definitions across files_, not just the type declaration in current file from which the fix is triggered.
2. If the trigger diagnostic span does not have a method level member declaration as it's containing symbol, we do not offer `ContainingMember` fix all option  (for example `make class static` code fix for [CA1822](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1822)).
3. Similarly, if the trigger diagnostic span does not have a type declaration as it's containing symbol, we do not offer `ContainingType` fix all option (for example `remove unnecessary usings` code fix for [IDE0005](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0005)).

## Added Tests

1. Unit tests: C# and VB unit tests tests for both `FixAllScope.ContainingMember` and `FixAllScope.ContainingType` in following buckets:
   1. Method body level code fix, so that both these fix all scopes are applicable for the code fix.
   4. Type declaration level code fix, so that only ContainingType fix all scope is applicable. We verify that ContainingMember fix all scope is not offered for such a code fix.
   5. File level code fix (remove unnecessary usings), so that neither of these fix all scopes are applicable. We verify that both these fix all scopes are not offered for such a code fix.
   6. FixAll tests for suppressions for both these scopes.
2. Integration tests for both `FixAllScope.ContainingMember` and `FixAllScope.ContainingType`.

mavasani commented on src/Features/Core/Portable/UnifiedSuggestions/UnifiedSuggestedActionsSource.cs:
```diff
@@ -210,6 +215,14 @@ static CodeFixGroupKey GetGroupKey(CodeFix fix)
             using var fixAllSuggestedActionsDisposer = ArrayBuilder<IUnifiedSuggestedAction>.GetInstance(out var fixAllSuggestedActions);
             foreach (var scope in supportedScopes)
             {
+                if (scope is FixAllScope.ContainingMember or FixAllScope.ContainingType)
```
This is the only real change in this file, rest all is making the entire call chain async. We bail out from showing ContainingMember or ContainingType fix all option if we cannot find a containing member or type respectively.

CyrusNajmabadi replied:
open question.  how does TS/F# play into this?  is finding the containing type/member part of some API they can imeplement?

mavasani replied:
That is outside the scope of this PR. If we want to add support for these scopes in TS/F#, we would need some way to be able to query them to get the containing member span for trigger position.

mavasani replied:
I actually went ahead and did this refactoring in this PR itself with https://github.com/dotnet/roslyn/pull/59989/commits/64baa64648a00220629e760d343a8b237728a9be. Other languages can now implement the `IFixAllSpanMappingService` language service to enable fix all for these new scopes.

jmarolf commented on src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1,5 +1,14 @@
+abstract Microsoft.CodeAnalysis.CodeFixes.FixAllContext.SpanBasedDiagnosticProvider.GetDocumentSpanDiagnosticsAsync(Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Text.TextSpan fixAllSpan, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Diagnostic>>
+Microsoft.CodeAnalysis.CodeFixes.DocumentBasedFixAllProvider.DocumentBasedFixAllProvider(System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CodeFixes.FixAllScope> supportedFixAllScopes) -> void
```
Are these public API additions intentional?

jmarolf replied:
I think we can duplicate some types here and make them internal.

333fred replied:
Yes, there's an issue template for it. Manish, you're one of the area owners so go ahead and file the issue and apply the `api-ready-for-review` tag.

mavasani replied:
@jmarolf Yes, these APIs are intentionally public. Otherwise, public clients such as MS.CA.Testing library, would be unable to create a FixAll provider that supports these new FixAll scopes. We can certainly keep them internal in this PR and have a separate PR to just make these APIs public.

mavasani replied:
https://github.com/dotnet/roslyn/issues/60029 tracks making APIs public. I'll switch the APIs to internal for this PR

mavasani replied:
I have internalized most of these APIs with https://github.com/dotnet/roslyn/pull/59989/commits/2df4e07f971dc80771b249de77e04c453eea0d9f. I still had to retain one public API. I will open an issue with the public API proposals.

jmarolf commented on src/EditorFeatures/Core/Shared/Extensions/TelemetryExtensions.cs:
```diff
@@ -40,7 +40,9 @@ public static short GetScopeIdForTelemetry(this FixAllScope scope)
                 FixAllScope.Project => 2,
                 FixAllScope.Solution => 3,
-                _ => 4,
+                FixAllScope.ContainingMember => 4,
```
We previously reported "Custom" scopes as `4` but will now report them as `6` with this change right? I assume we want to keep the telemetry reporting the same for custom scopes

mavasani replied:
Ah, thanks good catch. I'll make the change.

jmarolf commented on src/Features/Core/Portable/CodeFixes/CodeFixService.cs:
```diff
@@ -758,6 +760,17 @@ private async Task<IEnumerable<Diagnostic>> GetDocumentDiagnosticsAsync(Document
             return await diagnostics.ToDiagnosticsAsync(document.Project, cancellationToken).ConfigureAwait(false);
         }
 
+        private async Task<IEnumerable<Diagnostic>> GetDocumentSpanDiagnosticsAsync(Document document, TextSpan span, ImmutableHashSet<string>? diagnosticIds, bool includeSuppressedDiagnostics, CancellationToken cancellationToken)
```
Shouldn't this method be called by the one above and just pass in the span of the document?

mavasani replied:
No, the IDE DiagnosticAnalyzerService uses different diagnostic getters for span based diagnostic queries and full document diagnostic queries, each of which is optimized for span vs full-span scenarios. [LatestDiagnosticsForSpanGetter](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs,55) for span based queries and [IdeLatestDiagnosticGetter](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs,35) for full document/project/solution queries.

jmarolf commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DocumentBasedFixAllProvider.cs:
```diff
@@ -106,9 +115,26 @@ private static async Task<ImmutableArray<Diagnostic>> DetermineDiagnosticsAsync(
+                }
+                else
+                {
+                    var diagnostics = ImmutableArray<Diagnostic>.Empty;
```
Use a builder?

mavasani replied:
So, for 99% cases we expect a single FixAll span entry here. We get multiple FixAll spans within a document only for the case where a single named type has multiple partial definitions defined in the same file, which is almost never the case in real code bases. I don't think we need to optimize for this corner case.

CyrusNajmabadi replied:
my general concern is simply that the builder approach is just as easy to write, and works well for the 1 to many case.  I agree it's unlikely, but really rthe ImmutableArray.AddRange approach really stands out like a sort thumb as "this is just waiting for an unfortunate case.  Note: sam's TemporaryArray is great here.  just do:

```c#
using var diagnostics = TemporaryArray<Diagnostic>Empty;
```

As fast for the normal case (<4 items), and no pathlogical dropoff.

jmarolf commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.SpanBasedDiagnosticProvider.cs:
```diff
@@ -0,0 +1,41 @@
+    /// <summary>
+    /// Context for "Fix all occurrences" code fixes provided by a <see cref="FixAllProvider"/>.
+    /// </summary>
+    public partial class FixAllContext
```
Make Internal?

mavasani replied:
@jmarolf This type is already public. Note that this PR is not for FixAll support for code refactorings that we discussed last week. This PR is extending the existing FixAll feature and public APIs for diagnostics/code fixes to add new fix all scopes.

CyrusNajmabadi commented on src/Features/Core/Portable/FeaturesResources.resx:
```diff
@@ -737,6 +737,15 @@ Do you want to continue?</value>
   </data>
+  <data name="Selection" xml:space="preserve">
+    <value>Selection</value>
+  </data>
```
remove this one?

mavasani replied:
Ah, yes. Copy-paste is bad.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAll/FixAllState.cs:
```diff
@@ -28,24 +29,42 @@ internal sealed partial class FixAllState
+        /// <summary>
+        /// Optional fix all spans to be fixed within the document. Can be empty
+        /// if fixing the entire document, project or solution.
+        /// If non-empty, <see cref="Document"/> must be non-null.
```
'non-null' doesn't fit here.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -50,6 +54,48 @@ internal static class FixAllContextHelper
+                            foreach (var fixAllSpan in fixAllSpans)
+                            {
+                                documentDiagnostics = documentDiagnostics.AddRange(await fixAllContext.GetDocumentSpanDiagnosticsAsync(
+                                    document, fixAllSpan).ConfigureAwait(false));
```
can this helper just take in many spans?  woudl that be more efficient down below instead of calling multiple times into it?

mavasani replied:
As explained in https://github.com/dotnet/roslyn/pull/59989#discussion_r821315635, multiple fix all spans is going to be an extreme corner case here that we do not need to optimize for. For all practical purposes and performance measurements, we should just assume a single FixAll span within a document.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -50,6 +54,48 @@ internal static class FixAllContextHelper
+                        if (!fixAllSpans.IsEmpty)
+                        {
+                            // We are processing a span-based FixAllContext created for a partial declaration.
+                            var documentDiagnostics = ImmutableArray<Diagnostic>.Empty;
```
```suggestion
                            using var _ = ArrayBuilde<Diagnostic>.GetInstance(out var documentDiagnostics);
```

mavasani replied:
Same comment as above https://github.com/dotnet/roslyn/pull/59989#discussion_r821317353. We don't need to optimize for multiple fix all spans within a document case here.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -50,6 +54,48 @@ internal static class FixAllContextHelper
+                            // using the trigger span.
+                            var documentsAndSpans = await GetDocumentsAndSpansForContainingMemberOrTypeAsync(document,
+                                fixAllContext.Scope, triggerSpan.Value, fixAllContext.CancellationToken).ConfigureAwait(false);
+                            var unused = PooledDictionary<Document, ImmutableArray<Diagnostic>>.GetInstance(out var builder);
```
we have helpers already that go from an mutable dict builder with ArrayBuilder values, which can then be converted to an immutable/immutable collection.  The extension is called MultiAdd, and you can see the pattern in CodeFixService.cs i believe.

mavasani replied:
None of the existing overloads work for this case as we need to append array of diagnostics in each loop iteration. Also given we don't expect multiple fix all spans for each document in majority of cases, we should be fine here.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -137,8 +183,90 @@ public static string GetDefaultFixAllTitle(FixAllContext fixAllContext)
+            Contract.ThrowIfFalse(fixAllScope is FixAllScope.ContainingMember or FixAllScope.ContainingType);
+
+            var decl = await GetContainingMemberOrTypeDeclarationAsync(document, fixAllScope, triggerSpan, cancellationToken).ConfigureAwait(false);
+            if (decl == null)
```
when does this happen?  do we offer such a fixallscope if the user is not in sucha  scope?

mavasani replied:
This will happen for file level diagnostics, say unnecessary usings. We will not show such a fix scope to users for a code fix to such diagnostics. That filtering is done here: https://github.com/dotnet/roslyn/pull/59989#discussion_r820627128

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -137,8 +183,90 @@ public static string GetDefaultFixAllTitle(FixAllContext fixAllContext)
+            }
+            else
+            {
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```
this seems to be C#/VB specific, whcih concerns me.

mavasani replied:
Yes, that is intentionally so for this PR to keep it simple and limit the newly added scopes to just C#/VB for now. As mentioned in https://github.com/dotnet/roslyn/pull/59989#discussion_r821310498, we can absolutely enhance this in future to enable TS/F# to support these scopes, for which we would need to extract out all the code to map a trigger span to containing member/type span to some language service. I didn't want to add it here to avoid making this PR unnecessary bulkier.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.cs:
```diff
@@ -13,10 +13,18 @@ namespace Microsoft.CodeAnalysis.CodeFixes
 
-        protected SyntaxEditorBasedCodeFixProvider(bool supportsFixAll = true)
-            => _supportsFixAll = supportsFixAll;
+        protected SyntaxEditorBasedCodeFixProvider(bool supportsFixAll = true, ImmutableArray<FixAllScope> supportedFixAllScopes = default)
```
what are the features that pass something other than default here?

mavasani commented:
> Thinking in it more. Maybe change the name to something simple then doc which Scopes this is NOT called for. Then doc that this is a way for a language to customize any other scopes that are supported. ?

Thanks. Done with https://github.com/dotnet/roslyn/pull/59989/commits/de6c22d33b281ef333c941fa0c87fef589bc2f36

jmarolf commented on src/EditorFeatures/Core/Shared/Extensions/TelemetryExtensions.cs:
```diff
@@ -40,7 +40,10 @@ public static short GetScopeIdForTelemetry(this FixAllScope scope)
+                FixAllScope.Custom => 4,
+                FixAllScope.ContainingMember => 5,
+                FixAllScope.ContainingType => 6,
+                _ => 7,
```
hate to nit on this but before we reported `Custom` or `Unknown` as `4`. now we report `Custom` as 4 and `Unknown` as 7. I am fine taking this change but I just need to review out telemetry scripts and make sure I account for this new split. I'll reply back with a :+1: when I am sure we can handle this across past and future telemetry.

mavasani commented:
@jmarolf @CyrusNajmabadi any other feedback here? I have filed a separate issue for public API review, and will take care of any API changes as part of that work. Can you please review if this PR looks fine functionality wise to merge in?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Features/UnifiedSuggestions/UnifiedSuggestedActionsSource.cs:
```diff
@@ -207,9 +213,19 @@ static CodeFixGroupKey GetGroupKey(CodeFix fix)
             foreach (var scope in supportedScopes)
             {
+                if (scope is FixAllScope.ContainingMember or FixAllScope.ContainingType &&
+                    document.GetLanguageService<IFixAllSpanMappingService>() is IFixAllSpanMappingService spanMappingService)
```
interesting.  so if we have a scope of ContainingMember/Type and the language does *not* export sucha  service, then we proceed?  should we filter those out instead?

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DefaultFixAllProviderHelpers.cs:
```diff
@@ -74,5 +75,28 @@ internal static class DefaultFixAllProviderHelpers
+                return null;
+
+            var fixAllContexts = documentsAndSpansToFix.SelectAsArray(
+                (documentAndSpans) => fixAllContext.WithDocumentAndProject(documentAndSpans.Key, documentAndSpans.Key.Project)
```
asides: WithDoucmentAndProject feels somewhat weird given that a project is inferrable from a DOcument.  
also, i'm nto a fan of all the separate With'ers.  Could just have one and do `With(document: docAndSpan.Key, fixAllSpans: docAndSpan.Value)`

mavasani replied:
There are existing callsites for this API that pass a null document and non-null project: https://github.com/dotnet/roslyn/blob/7e835c8c324e294b467ba69648a2049e3acbcc6f/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DefaultFixAllProviderHelpers.cs#L51 and https://github.com/dotnet/roslyn/blob/7e835c8c324e294b467ba69648a2049e3acbcc6f/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DefaultFixAllProviderHelpers.cs#L75

> i'm nto a fan of all the separate With'ers

I agree, but going to leave this as is to avoid unrelated stylistic API changes in this PR.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.SpanBasedDiagnosticProvider.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <see cref="GetDocumentSpanDiagnosticsAsync(Document, TextSpan, CancellationToken)"/> method to compute
+        /// the diagnostics only for the given 'fixAllSpan' upfront.
+        /// </summary>
+        internal abstract class SpanBasedDiagnosticProvider : DiagnosticProvider
```
what is slightly unclear to me is who is intended to subclass this.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -88,12 +96,50 @@ public partial class FixAllContext
+        {
+            if (scope is FixAllScope.ContainingMember or FixAllScope.ContainingType)
+            {
+                throw new ArgumentException(WorkspacesResources.Fixallscope_containingtype_and_fixallscope_containingmember_are_not_supported_with_this_constructor,
```
```suggestion
                throw new ArgumentException(WorkspacesResources.FixAllScope_ContainingType_and_FixAllScope_ContainingMember_are_not_supported_with_this_constructor,
```

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -163,23 +216,55 @@ public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Docume
                 if (diagnostics != null)
                 {
-                    return diagnostics.Where(d => d != null && diagnosticIds.Contains(d.Id)).ToImmutableArray();
+                    return diagnostics.Where(d => d != null && diagnosticIds.Contains(d.Id)
```
```suggestion
                    return diagnostics.WhereAsArray(d => d != null && diagnosticIds.Contains(d.Id)
```

mavasani replied:
`WhereAsArray` is an extension method on `ImmutableArray<T>`, but not available on `IEnumerable<T>`

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -163,23 +216,55 @@ public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Docume
+            }
+            else
+            {
+                getDiagnosticsTask = State.DiagnosticProvider.GetDocumentDiagnosticsAsync(document, this.CancellationToken);
```
is there somethign that post-filters the diagnostics from this to the passed in filterspan?  i'm not seeing it.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -163,23 +216,55 @@ public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Docume
+        private static async Task<ImmutableArray<Diagnostic>> GetFilteredDiagnosticsAsync(
+            Task<IEnumerable<Diagnostic>> getDiagnosticsTask,
+            ImmutableHashSet<string> diagnosticIds,
+            TextSpan? filterSpan)
```
is this only called from one place?  consider making nested helper method.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -163,23 +216,55 @@ public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Docume
+            ImmutableHashSet<string> diagnosticIds,
+            TextSpan? filterSpan)
         {
             if (getDiagnosticsTask != null)
```
can' this actually ever be null?  i'm not seeing how.

CyrusNajmabadi replied:
bleagh.  i hate thsi.  but i'm ok with preserving it....

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs:
```diff
@@ -59,6 +59,38 @@ public static FixAllProvider Create(Func<FixAllContext, Document, ImmutableArray
             return new CallbackDocumentBasedFixAllProvider(fixAllAsync);
```
consider having this method defer to the one below it.

mavasani replied:
That method requires supported fix all scopes, and GetDefaultSupportedFixAllScopes is an instance method on FixAllProvider. `CallbackDocumentBasedFixAllProvider` has two ctors, which call into different base type ctors, so these are different code paths.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs:
```diff
@@ -68,6 +100,14 @@ public CallbackDocumentBasedFixAllProvider(Func<FixAllContext, Document, Immutab
                 _fixAllAsync = fixAllAsync;
```
my *string* preference is that constructors chain horizontally in a type (e.g. use `this`), then a single constructor uses `base` to walk up the inheritance hierarchy.  That way any special logic (like assigning fields) is always only in one constructor.

mavasani replied:
Ideally, yes. But we don't have access to `GetDefaultSupportedFixAllScopes` API here, so we need this ctor to call into the parameterless public ctor of `DocumentBasedFixAllProvider`

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllScope.cs:
```diff
@@ -12,6 +12,8 @@ public enum FixAllScope
-        Custom
+        Custom,
+        ContainingMember,
+        ContainingType,
```
public API.  please consider doc'ing.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1,5 +1,9 @@
+Microsoft.CodeAnalysis.CodeFixes.FixAllScope.ContainingType = 5 -> Microsoft.CodeAnalysis.CodeFixes.FixAllScope
 Microsoft.CodeAnalysis.Editing.SyntaxEditor.SyntaxEditor(Microsoft.CodeAnalysis.SyntaxNode root, Microsoft.CodeAnalysis.Host.HostWorkspaceServices services) -> void
 *REMOVED*static Microsoft.CodeAnalysis.Editing.SyntaxGenerator.DefaultRemoveOptions -> Microsoft.CodeAnalysis.SyntaxRemoveOptions
+static Microsoft.CodeAnalysis.CodeFixes.FixAllProvider.Create(System.Func<Microsoft.CodeAnalysis.CodeFixes.FixAllContext, Microsoft.CodeAnalysis.Document, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostic>, System.Threading.Tasks.Task<Microsoft.CodeAnalysis.Document>> fixAllAsync, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CodeFixes.FixAllScope> ...
```
we could consider not making this public for now, right? (i'm fine either way).

mavasani replied:
`FixAllProvider.Create` is used by the CodeStyle layer to create custom `DocumentBasedFixAllProvider`, and it only has access to public APIs in the Workspace layer. So we need this API to be public if we want CodeStyle NuGet based code fixes to show the new ContainingMember and ContainingType FixAll scopes.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/WorkspacesResources.resx:
```diff
@@ -576,4 +576,10 @@
     <value>Predefined conversion from {0} to {1}.</value>
   </data>
+  <data name="Fixallscope_containingtype_and_fixallscope_containingmember_are_not_supported_with_this_constructor" xml:space="preserve">
+    <value>'FixAllScope.ContainingType' and 'FixAllScope.ContainingMember' are not supported with this constructor</value>
```
so the loc strings should match the actual english content as much as possible.  so that means matching cases as well.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAll/FixAllState.cs:
```diff
@@ -28,24 +29,42 @@ internal sealed partial class FixAllState
+        /// <summary>
+        /// Optional fix all spans to be fixed within the document. Can be empty
+        /// if fixing the entire document, project or solution.
+        /// If non-empty, <see cref="Document"/> is guaranteed to be not null.
```
```suggestion
        /// If non-empty, <see cref="Document"/> is guaranteed to be not <see langword="default"/>.
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAll/FixAllState.cs:
```diff
@@ -64,29 +83,37 @@ public FixAllState WithCodeActionEquivalenceKey(string? codeActionEquivalenceKey
             Optional<FixAllScope> scope = default,
-            Optional<string?> codeActionEquivalenceKey = default)
+            Optional<string?> codeActionEquivalenceKey = default,
+            Optional<ImmutableArray<TextSpan>> fixAllSpans = default)
```
happy there is a 'With' here :)  i wish that other type was written the same way.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -50,6 +54,49 @@ internal static class FixAllContextHelper
                     break;
 
+                case FixAllScope.ContainingMember or FixAllScope.ContainingType:
+                    if (document != null && !await document.IsGeneratedCodeAsync(cancellationToken).ConfigureAwait(false))
```
doc why the IsGeneratedCode check.  i'm not getting why that is here.

mavasani replied:
This check already existed for `FixAllScope.Document` above, so I just retained it to match behavior for partial type declarations in generated code: https://github.com/dotnet/roslyn/blob/7bb079e5ceb700d1772b88f7f1b798807fe31805/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs#L44-L45

I believe @sharwell added this check to ensure that FixAll operations don't end up changing generated code.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -50,6 +54,49 @@ internal static class FixAllContextHelper
+                                    var documentDiagnostics = await fixAllContext.GetDocumentSpanDiagnosticsAsync(documentToFix, span).ConfigureAwait(false);
+                                    if (builder.TryGetValue(documentToFix, out var existingDiagnostics))
+                                    {
+                                        documentDiagnostics = existingDiagnostics.AddRange(documentDiagnostics);
```
i don't get what's happening here.  if we just computed the doc span diagnostics 3 lines above... why are we adding other diagnostics to it?

mavasani replied:
Should hopefully be much easier to follow after the refactoring.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.cs:
```diff
@@ -13,10 +13,16 @@ namespace Microsoft.CodeAnalysis.CodeFixes
-            => _supportsFixAll = supportsFixAll;
+        {
+            _supportsFixAll = supportsFixAll;
+        }
```
unnecessary change.  but fine if you prefer the block form.

mavasani replied:
I had edited the ctor to add another statement and then reverted, but forgot to switch the code back to expression body form.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,91 @@
+    internal abstract class AbstractFixAllSpanMappingService : IFixAllSpanMappingService
+    {
+        public async Task<ImmutableDictionary<Document, ImmutableArray<TextSpan>>> GetFixAllSpansAsync(
+           Document document, TextSpan triggerSpan, FixAllScope fixAllScope, CancellationToken cancellationToken)
```
incorrect indentation.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,91 @@
+            if (fixAllScope == FixAllScope.ContainingMember)
+            {
+                return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty
+                    .SetItem(document, ImmutableArray.Create(decl.FullSpan));
```
nit:

```suggestion
                    .Add(document, ImmutableArray.Create(decl.FullSpan));
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,91 @@
+                            continue;
+                        var root = await syntaxRef.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+                        var partialDeclSpan = syntaxFacts.GetContainingTypeDeclaration(root, syntaxRef.Span.Start)!.FullSpan;
+                        builder.MultiAdd(documentForLocation, partialDeclSpan);
```
multi-add works with immutable-array?

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,91 @@
+                        builder.MultiAdd(documentForLocation, partialDeclSpan);
+                    }
+
+                    return builder.ToImmutableDictionary();
```
```suggestion
                    return builder.ToImmutableDictionary(kvp => kcp.Key, kvp => kvp.Value.ToImmutable());
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,91 @@
+            if (fixAllScope == FixAllScope.ContainingMember && !syntaxFacts.IsMethodLevelMember(startContainer))
+                return null;
+
+            if (triggerSpan.IsEmpty || startContainer == null)
```
1. can you split this heck, it's pretty confusing.  especially "if start is null, return start".

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/WorkspaceExtensionsResources.resx:
```diff
@@ -129,6 +129,12 @@
   <data name="Fix_all_0_in_Solution" xml:space="preserve">
     <value>Fix all '{0}' in Solution</value>
   </data>
+  <data name="Fix_all_0_in_Containing_Member" xml:space="preserve">
```
string doesn't match text (casing).

CyrusNajmabadi reviewed (changes requested):
overall strongly like this.  but a few things i def would like to see a little attention on (And clarity).  thanks!

mavasani commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DefaultFixAllProviderHelpers.cs:
```diff
@@ -74,5 +75,28 @@ internal static class DefaultFixAllProviderHelpers
+                return null;
+
+            var fixAllContexts = documentsAndSpansToFix.SelectAsArray(
+                (documentAndSpans) => fixAllContext.WithDocumentAndProject(documentAndSpans.Key, documentAndSpans.Key.Project)
```
A code style analyzer or refactoring to suggest this change would be good :-)

mavasani commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -163,23 +216,55 @@ public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Docume
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (this.Project.Language != document.Project.Language)
```
Few things to note:
1. This is a check for a public API input, that mimics the existing checks for other similar public APIs in this type: https://github.com/dotnet/roslyn/blob/7bb079e5ceb700d1772b88f7f1b798807fe31805/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs#L150-L163
2. This public API is used by custom FixAllProviders to compute diagnostics to fix in any given document that is part of the FixAllScope that the provider has been asked to fix. For a multi-language solution, a FixAll in solution will span all documents in the solution, but current FixAll implementation and design only supports fixing the projects whose language match the FixAll trigger document language - FixAll in C# document will not fix similar VB documents. It is possible that custom FixAllProviders call this API for all documents in the solution, and the API appropriately returns empty set of diagnostics to fix in documents of different language that the trigger document language.

mavasani commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -163,23 +216,55 @@ public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Docume
 
-        private static async Task<ImmutableArray<Diagnostic>> GetFilteredDiagnosticsAsync(Task<IEnumerable<Diagnostic>> getDiagnosticsTask, ImmutableHashSet<string> diagnosticIds)
+        private static async Task<ImmutableArray<Diagnostic>> GetFilteredDiagnosticsAsync(
+            Task<IEnumerable<Diagnostic>> getDiagnosticsTask,
```
> really don't love seeing tasks of streaming work.

Existing public API :-( https://github.com/dotnet/roslyn/blob/7bb079e5ceb700d1772b88f7f1b798807fe31805/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.DiagnosticProvider.cs#L23-L28

mavasani commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/FixAllContextHelper.cs:
```diff
@@ -50,6 +54,49 @@ internal static class FixAllContextHelper
+                            return ImmutableDictionary<Document, ImmutableArray<Diagnostic>>.Empty.SetItem(document, documentDiagnostics);
+                        }
+                        else if (triggerSpan.HasValue &&
+                            document.GetLanguageService<IFixAllSpanMappingService>() is IFixAllSpanMappingService spanMappingService)
```
So this source file is in the shared layer used by CodeStyle and Workspaces layer, so we cannot access the internal APIs to get to `FixAllContext.State.FixAllSpans` or `FixAllContext.TriggerSpan`, we have to pass the `triggerSpan` and `fixAllSpans` as parameters to this method. The reason we need both these parameters is the way `FixAllContext.State` is built up by the `DocumentBasedFixAllProvider` - we start with an instance of `FixAllContext` that just has the original input `triggerSpan` and not yet computed `FixAllSpans`, and later on the `DocumentBasedFixAllProvider` mutates the `FixAllContext.State` to populate it with computed `FixAllSpans` from the original trigger span.
The `BatchFixAllProvider` invokes this shared helper method with the original `FixAllContext` (i.e. with a trigger span, but not yet computed FixAllSpans), while the `DocumentBasedFixAllProvider` invokes this shared helper with the mutated `FixAllContext`, which already has the computed `FixAllSpans`. The if condition in this else if and the prior if check for both these code paths and has appropriate code to compute the final diagnostics to fix. Let me refactor/share/doc this code a bit more to help clarify.

mavasani commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,91 @@
+                        if (documentForLocation == null)
+                            continue;
+                        var root = await syntaxRef.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+                        var partialDeclSpan = syntaxFacts.GetContainingTypeDeclaration(root, syntaxRef.Span.Start)!.FullSpan;
```
I had the code as per your suggestion earlier, but VB has the weird syntax nodes where the SourceSpan of the syntax ref is only the ClassStatementSyntax, not the containing ClassBlockSyntax. This code pattern handles both C# and VB correctly. We also have a unit test that verifies the C# and VB partial decl case which guards a regression here.

mavasani commented:
@CyrusNajmabadi Thank you for the review! I believe I have addressed all your concerns, PTAL.

mavasani commented:
@CyrusNajmabadi @jmarolf Any further feedback?

mavasani commented:
@CyrusNajmabadi @jmarolf @333fred I have addressed all the API review feedback from https://github.com/dotnet/roslyn/issues/60029#issuecomment-1071458718

1. Added tests and support for top level statements with https://github.com/dotnet/roslyn/pull/59989/commits/9f346b516da46699fe2a523f0a42e892195632ce. This was a good catch, before this commit the new fix all scopes wouldn't have shown up within top level statements.
2. Renamed `triggerSpan` to `diagnosticSpan` with https://github.com/dotnet/roslyn/pull/59989/commits/85f433755a1f1c3298167032dd7ce0c194c8c7df
3. Switched `GetDocumentSpanDiagnosticsAsync` to internal with https://github.com/dotnet/roslyn/pull/59989/commits/f7ac5cd434c92ba97b9300aabf0d81f01ae6cb25
4. Made rest of the APIs approved in API review public with https://github.com/dotnet/roslyn/pull/59989/commits/516a15b5a51cc8484802a2a1400bb1f412e541bb

mavasani commented:
Ping @CyrusNajmabadi @jmarolf - hopefully should be a quick review pass

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContextHelper.cs:
```diff
@@ -0,0 +1,167 @@
+                        //  2. Using the original 'FixAllContext' which has the 'triggerSpan' and 'document' for the original code fix, but
+                        //     the 'fixAllSpans' have not yet been computed. For this scenario, we use the 'IFixAllSpanMappingService' to
+                        //     map the triggerSpan and fixAllScope to documents and spans to fix for the containing member or
+                        //     type declaration (and its partials), and then return the diagnostics for each of the documents and fixAllSpans.
```
this is interesting and thanks for the info.  but could you clarify why some paths have a `fully populated 'FixAllContext', which has already computed, non-empty 'fixAllSpans"`, while some don't?  It's genuinely not clear to me how that would happen.

mavasani replied:
Thanks! So, there seemed to be couple of different code paths in to reach this helper and at some place the FixAllSpans were lazily computed and at some they were precomputed and passed around. I have cleaned this up by completely removing `FixAllState.FixAllSpans` property and instead we now always lazily compute these spans right when they are required to compute diagnostics to fix.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContextHelper.cs:
```diff
@@ -0,0 +1,167 @@
+                            builder.Add(document, arrayBuilder);
+                        }
+
+                        arrayBuilder.AddRange(documentDiagnostics);
```
i believe there's already a multi-add helper for this so you don't need to do the ArrayBuilder stuff above.

mavasani replied:
`MultiAdd` extension only works if the value being added is a single diagnostic. I have added a new extension `MultiAddRange` and used it.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContextHelper.cs:
```diff
@@ -0,0 +1,167 @@
+                FixAllContext fixAllContext,
+                IEnumerable<KeyValuePair<Document, ImmutableArray<TextSpan>>> documentsAndSpans)
+            {
+                var builder = PooledDictionary<Document, ArrayBuilder<Diagnostic>>.GetInstance();
```
oh.  this was a move.  n/m. you don't have to change anything thnat was a direct move if you don't want to.

mavasani replied:
Yes, just a code move so will retain this as-is.

CyrusNajmabadi commented on src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/CSharpCodeFixVerifier`2+Test.cs:
```diff
@@ -122,7 +122,7 @@ protected override CodeFixContext CreateCodeFixContext(Document document, TextSp
                 => new(new FixAllState(
                     fixAllProvider: null,
-                    triggerSpan: null,
+                    diagnosticSpan: null,
```
much clearer.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -107,13 +107,14 @@ public partial class FixAllContext
+        /// is not required and other constructor which does not take a diagnostic span can be used instead.
         /// </summary>
         /// <param name="document">Document within which fix all occurrences was triggered.</param>
+        /// <param name="diagnosticSpan">Span for the diagnostic for which fix all occurrences was triggered.</param>
```
Dumb question.  Should we just include the original diagnostic itself?  That seems 'reasonable' to me in that you started with a diagnostic, and you're asking to "fix all" of the diagnostics like it in some scope.

I basically worry that we might need some of the other data on it in the future (like teh properties/etc.) so we might just include it now for completeness.

mavasani replied:
I had started with attempting to use diagnostic as the input, but the CodeFixContext exposes a single span and multiple possible diagnostics of the same ID for the span: https://github.com/dotnet/roslyn/blob/53468b541401e6d1fe358cc4997fcabc37ea987c/src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs#L32-L41

Additionally, the `RegisterCodeFix` method has overloads to register a single code action fixing multiple of these diagnostics: https://github.com/dotnet/roslyn/blob/53468b541401e6d1fe358cc4997fcabc37ea987c/src/Workspaces/Core/Portable/CodeFixes/CodeFixContext.cs#L146-L166

So it seemed more logical to stick to the diagnostic span as the input (which is actually `CodeFixContext.Span` to be honest).

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DocumentBasedFixAllProvider.cs:
```diff
@@ -123,14 +123,14 @@ private static async Task<ImmutableArray<Diagnostic>> DetermineDiagnosticsAsync(
                     }
 
-                    return diagnostics;
+                    return diagnostics.ToImmutable();
```
does this code run oop?  if so, consider getting doc diagnostics in parallel.

mavasani replied:
This doesn't run in OOP

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpFixAllSpanMappingService.cs:
```diff
@@ -19,5 +27,33 @@ internal sealed class CSharpFixAllSpanMappingService : AbstractFixAllSpanMapping
+            var firstNonGlobalStatement = root.ChildNodes().FirstOrDefault(n => n is not GlobalStatementSyntax);
+            if (firstNonGlobalStatement is not null)
+            {
+                // Bail out for error case where a non-global statement precedes global statements.
```
mention that this could would be in error, so it's ok to have this strategy.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpFixAllSpanMappingService.cs:
```diff
@@ -19,5 +27,33 @@ internal sealed class CSharpFixAllSpanMappingService : AbstractFixAllSpanMapping
+            {
+                // Bail out for error case where a non-global statement precedes global statements.
+                if (root.FullSpan.Start == firstNonGlobalStatement.FullSpan.Start)
+                    return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty;
```
i don't think this check is right.  for example, you could have 

```
using X;

class C { }

if (x) { ... }
```

What i think you simply want to check is if there is a non-global statement that precedes a global statement.

mavasani replied:
Done. thanks

CyrusNajmabadi commented:
Done with pass.

mavasani commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DefaultFixAllProviderHelpers.cs:
```diff
@@ -75,28 +73,5 @@ internal static class DefaultFixAllProviderHelpers
                 sortedProjects.SelectAsArray(p => fixAllContext.WithScope(FixAllScope.Project).WithDocumentAndProject(document: null, p)));
         }
-
-        private static async Task<Solution?> GetContainingMemberOrTypeFixesAsync(FixAllContext fixAllContext, FixAllContexts fixAllContextsAsync)
```
This code path was precomputing and eagerly populating FixAllSpans.

mavasani commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -123,8 +123,7 @@ public partial class FixAllContext
         /// <param name="cancellationToken">Cancellation token for fix all computation.</param>
-        /// <remarks>TODO: Make API public (tracked by https://github.com/dotnet/roslyn/issues/60029)</remarks>
-        internal FixAllContext(
+        public FixAllContext(
```
I don't think we should - hosts which don't want to support the newly added FixAllScopes don't need to switch over to the new constructor.

mavasani commented:
Thanks @CyrusNajmabadi - I believe I have addressed all the comments and cleaned up the code quite a bit more.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Features/UnifiedSuggestions/UnifiedSuggestedActionsSource.cs:
```diff
@@ -207,9 +213,26 @@ static CodeFixGroupKey GetGroupKey(CodeFix fix)
+                {
+                    // Skip showing ContainingMember and ContainingType FixAll scopes if the language
+                    // does not implement 'IFixAllSpanMappingService' langauge service or
+                    // we have no mapped FixAll spans to fix.
```
this is a bit odd right?  other languages aren't going to add those FixAllScopes... so how would we get here?  we'd only get these scopes for C#/VB fixers taht update to use this, which means tehre is a mapping service...

mavasani replied:
This is a defensive check in case a third-party code fix provider for a different language returns the new FixAllScopes for `FixAllProvider.GetSupportedFixAllScopes`, but the language itself does not support/implement these scopes.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs:
```diff
@@ -27,6 +27,10 @@ private BatchFixAllProvider()
 
+        public override IEnumerable<FixAllScope> GetSupportedFixAllScopes()
+            => ImmutableArray.Create(FixAllScope.Document, FixAllScope.Project,
+                FixAllScope.Solution, FixAllScope.ContainingMember, FixAllScope.ContainingType);
```
oh... i see.  you're making it so that anyone that uses BatchFixAllProvider.Instance now is opted into this... hrmmmm

that's potentialy more concerning.  though i guess we can do this and then filter this out later when we process the diagnostics.  i'm not a huge fan though as i think it might be better to keep old behavior and have people opt in... 

but i do get that this is beneficial for all the C#/VB fixers out there that won't update hat could use this :-/

mavasani replied:
> i see. you're making it so that anyone that uses BatchFixAllProvider.Instance now is opted into this

Yes, this is something Sam and I discussed and decided to change.

> that's potentialy more concerning

I don't think so. Batch fixer works by merging code actions for diagnostics to fixed, and if it works fine for all diagnostics in a document, it should work fine for subset of those in a document as well.

> i'm not a huge fan though as i think it might be better to keep old behavior and have people opt in...

There is no way to opt-in. The BatchFixer is a complete FixAllProvider implementation. We can potentially have a new BatchFixer instance that supports additional modes, but in reality that would mean none of the existing code fixers will ever get updated to support new scopes. I would agree with your concern if this was done for FixAll for code refactorings, as we do not have a good way to guarantee that if it works on a document, it will work on a span. But I think making this assumption for code fixers is fine.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DocumentBasedFixAllProvider.cs:
```diff
@@ -106,9 +115,21 @@ private static async Task<ImmutableArray<Diagnostic>> DetermineDiagnosticsAsync(
+                else
+                {
+                    return await fixAllContext.GetDocumentDiagnosticsAsync(document).ConfigureAwait(false);
+                }
```
basically, the original code looks correct to me :)   it feels like this shoudl move inwards into the helper hte context already exposes.

mavasani replied:
Alright, so seems there was lot of wonkiness in the DocumentBasedFixAllProvider code. I have cleaned this all up to directly use the FixAllContextHelper to fetch the per-document diagnostics, and pass the map down to `GetFixedDocumentsAsync` helper. Hopefully, this makes everything much more clear.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.SpanBasedDiagnosticProvider.cs:
```diff
@@ -0,0 +1,41 @@
+            /// <summary>
+            /// Gets all the diagnostics to fix for the given <paramref name="fixAllSpan"/> in the given <paramref name="document"/> in a <see cref="FixAllContext"/>.
+            /// </summary>
+            public abstract Task<IEnumerable<Diagnostic>> GetDocumentSpanDiagnosticsAsync(Document document, TextSpan fixAllSpan, CancellationToken cancellationToken);
```
nit: consider returning ImmutableArray.

mavasani replied:
I have tried to be consistent with the existing public APIs in the base type: https://github.com/dotnet/roslyn/blob/ac603f8bdde93a35e2e4b9ef95a4674b561ab433/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.DiagnosticProvider.cs#L23-L39

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs:
```diff
@@ -88,8 +96,45 @@ public partial class FixAllContext
             IEnumerable<string> diagnosticIds,
             DiagnosticProvider fixAllDiagnosticProvider,
             CancellationToken cancellationToken)
+            : this(document, diagnosticSpan: null, codeFixProvider, scope,
```
can we add this to teh *roslyn* banned API list?  taht way we make sure that all our code calls teh new constructor?  i get the benefit of having this still in teh public API.  but we want to be careful we're not accidentally using this.

mavasani replied:
Sure, I'll do that.

mavasani replied:
Turns out both the existing public constructors for FixAllContext are already part of the banned API list, recommendation is to use the internal constructor overload:
https://github.com/dotnet/roslyn/blob/ac603f8bdde93a35e2e4b9ef95a4674b561ab433/eng/config/BannedSymbols.txt#L9-L10

I'll add even the new public API to this file.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContextHelper.cs:
```diff
@@ -0,0 +1,145 @@
+    {
+        public static async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixAsync(
+            FixAllContext fixAllContext)
+        {
```
so why is this not just an internal helper on FixAllContext?

mavasani replied:
It's used at couple of places - in `DocumentBasedFixAllProvider` and `FixAllContext.DiagnosticProvider`

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,61 @@
+                // C# compiler requires all global statements to preceed non-global statements.
+                var globalStatements = root.ChildNodes().OfType<GlobalStatementSyntax>();
+                if (globalStatements.Any(g => firstNonGlobalStatement.SpanStart < g.SpanStart))
+                    return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty;
```
this doesn't sppear fixed.  your `root.ChildNodes().FirstOrDefault(n => n is not GlobalStatementSyntax)` will return things like a `using directive`.  That using directive is allowed to come before a normal `global statement`.  but this will return empty here.

mavasani replied:
Thanks - I updated the check to now use [SyntaxFacts.IsNamespaceMemberDeclaration](https://sourceroslyn.io/Microsoft.CodeAnalysis.CSharp/R/9217de0879160b70.html), which matches the logic used by the language parser to report this error: https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs#L604-L628

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DocumentBasedFixAllProvider.cs:
```diff
@@ -30,8 +30,16 @@ namespace Microsoft.CodeAnalysis.CodeFixes
+
         protected DocumentBasedFixAllProvider()
         {
+            _supportedFixAllScopes = base.GetSupportedFixAllScopes().ToImmutableArray();
```
fwiw, this feels pretty wonky.  can we not hav FixAllProvider just make a protected internal static with the default `ImmutableArray.Create(FixAllScope.Document, FixAllScope.Project, FixAllScope.Solution)`?  the base woudl just return that.  here, we'd just defer to our other constructor, passing that along.

mavasani replied:
Done.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,96 @@
+                if (symbol?.DeclaringSyntaxReferences.Length > 1)
+                {
+                    var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                    var builder = PooledDictionary<Document, ArrayBuilder<TextSpan>>.GetInstance();
```
```suggestion
                    using var = PooledDictionary<Document, ArrayBuilder<TextSpan>>.GetInstance(out var builder);
```

mavasani replied:
PooledDictionary only supports `ToImmutableMultiDictionaryAndFree`, so I ended up using this pattern.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,96 @@
+                }
+                else
+                {
+                    return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty
+                        .SetItem(document, ImmutableArray.Create(decl.FullSpan));
```
```suggestion
                    return ImmutableDictionary.create(document, ImmutableArray.Create(decl.FullSpan));
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,96 @@
+
+            if (fixAllScope == FixAllScope.ContainingMember)
+            {
+                return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty
+                    .Add(document, ImmutableArray.Create(decl.FullSpan));
```
```suggestion
                return ImmutableDictionary.Create(document, ImmutableArray.Create(decl.FullSpan));
```

probably applies to other code in this PR :)

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,96 @@
+            if (fixAllScope == FixAllScope.ContainingMember && !syntaxFacts.IsMethodLevelMember(startContainer))
+                return null;
+
+            if (span.IsEmpty)
```
when is span empty?

mavasani replied:
Diagnostic with zero span size and non-zero start?

CyrusNajmabadi reviewed (changes requested):
done iwth pass.  almost all nits.  did see a concern with global statements.

mavasani commented:
Thanks @CyrusNajmabadi - feedback addressed

CyrusNajmabadi commented on src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContextHelper.cs:
```diff
@@ -0,0 +1,145 @@
+                    }
+                }
+
+                return builder.ToImmutableMultiDictionaryAndFree();
```
does this free the inner ArrayBuilder?

mavasani replied:
Yes, it does.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,61 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
```
remove.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,61 @@
+        {
+            Contract.ThrowIfFalse(fixAllScope is FixAllScope.ContainingMember or FixAllScope.ContainingType);
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```
```suggestion
            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,96 @@
+
+            if (fixAllScope == FixAllScope.ContainingMember)
+            {
+                return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty
```
```suggestion
                return ImmutableDictionary.Create(
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/FixAllSpanMappingService/AbstractFixAllSpanMappingService.cs:
```diff
@@ -0,0 +1,96 @@
+                }
+                else
+                {
+                    return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty
+                        .Add(document, ImmutableArray.Create(decl.FullSpan));
```
```suggestion
                    return ImmutableDictionary.Create(document, ImmutableArray.Create(decl.FullSpan));
```
